pydantic_ai_slim/pydantic_ai/agent.py

```diff
@@ -556,17 +568,29 @@ def system_prompt(
         def simple_system_prompt() -> str:
             return 'foobar'
 
-        @agent.system_prompt
+        @agent.system_prompt(dynamic=True)
         async def async_system_prompt(ctx: RunContext[str]) -> str:
             return f'{ctx.deps} is the best'
-
-        result = agent.run_sync('foobar', deps='spam')
-        print(result.data)
-        #> success (no tool calls)
         ```
         """
-        self._system_prompt_functions.append(_system_prompt.SystemPromptRunner(func))
-        return func
+        if func is None:
+
+            def decorator(
+                func_: _system_prompt.SystemPromptFunc[AgentDeps],
+            ) -> _system_prompt.SystemPromptFunc[AgentDeps]:
+                runner = _system_prompt.SystemPromptRunner(func_, dynamic=dynamic)
+                self._system_prompt_functions.append(runner)
+                if dynamic:
+                    self._system_prompt_dynamic_functions[func_.__qualname__] = runner
```

Comment:
I'm pretty sure this can't happen with the current overrides, better to have:

```suggestion
                assert not dynamic, "dynamic can't be True in this case"
                runner = _system_prompt.SystemPromptRunner(func_, dynamic=dynamic)
                self._system_prompt_functions.append(runner)
```